gprofiler/heartbeat.py

```diff
@@ -0,0 +1,627 @@
+                self.stop_event.wait(self.heartbeat_interval)
+    
+    def _stop_current_profiler(self):
+        """Stop the currently running profiler"""
+        if self.current_gprofiler:
+            logger.info("STOPPING current gProfiler instance...")
+            try:
+                self.current_gprofiler.stop()  # This sets the stop_event!
+                logger.info("Successfully called gprofiler.stop()")
+            except Exception as e:
+                # TODO: This is a huge leak, report it  
+                logger.error(f"Error stopping gProfiler: {e}")
+            
+            # ALWAYS cleanup subprocesses regardless of stop() success/failure
+            try:
+                logger.info("Starting comprehensive cleanup after heartbeat stop...")
+                self.current_gprofiler.maybe_cleanup_subprocesses()
+                logger.info("Comprehensive cleanup completed")
+            except Exception as cleanup_error:
+                # Cleanup errors are non-fatal - log and continue
+                logger.info(f"Cleanup completed with minor errors (expected during stop): {cleanup_error}")
+            
+            # Always clear the reference
+            self.current_gprofiler = None
+        
+        if self.current_thread and self.current_thread.is_alive():
+            # No need to actively kill the thread, the self.current_gprofiler.stop() already handles it using events
+            logger.info("Waiting for profiler thread to finish...")
+            self.current_thread.join(timeout=10)
+            self.current_thread = None
+    
+    def _start_new_profiler(self, profiling_command: Dict[str, Any], command_id: str):
+        """Start a new profiler with the given configuration"""
+        try:
+            # Import here to avoid circular imports
+            from gprofiler.main import DEFAULT_PROFILING_DURATION
+            
+            # Create modified args for the new profiler
+            new_args = self._create_profiler_args(profiling_command)
+            
+            # Create new GProfiler instance
+            self.current_gprofiler = self._create_gprofiler_instance(new_args)
+            
+            # Start profiler in a separate thread
+            self.current_thread = threading.Thread(
+                target=self._run_profiler,
+                args=(
+                    self.current_gprofiler,
+                    new_args.continuous,
+                    getattr(new_args, "duration", DEFAULT_PROFILING_DURATION),
+                    command_id,
+                ),
+                daemon=True
+            )
+            self.current_thread.start()
+            
+            logger.info(f"Started new gProfiler instance with command ID: {command_id}")
+            
+        except Exception as e:
+            logger.error(f"Failed to start new profiler: {e}", exc_info=True)
+            # Report failure to the server
+            self.heartbeat_client.send_command_completion(
```

Comment:
This code might send the duplicated command completion message when exception is raised. Please remove it if it's right.